File: super_camera/ios/Classes/SuperCameraPlugin.m
#import "SuperCameraPlugin.h"

@interface Pair : NSObject
@property(readwrite) FlutterMethodChannel *channel;
@property(readwrite) CameraController *controller;
@end

@implementation Pair
@end

@interface SuperCameraPlugin ()
@property NSMutableArray<Pair *> *controllers;
@property id<FlutterPluginRegistrar> registrar;
@end

@implementation SuperCameraPlugin
static NSString *const kPluginChannelName = @"bmparr2450.plugins/super_camera";

- (instancetype)initWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
  self = [super init];
  if (self) {
    _controllers = [NSMutableArray new];
    _registrar = registrar;
  }
  return self;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:kPluginChannelName
            binaryMessenger:[registrar messenger]];
  
  SuperCameraPlugin* instance = [[SuperCameraPlugin alloc] initWithRegistrar:registrar];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"Camera#availableCameras" isEqualToString:call.method]) {
    result([CameraController availableCameras]);
  } else if ([@"Camera#createCameraController" isEqualToString:call.method]) {
    [self createCameraController:call];
    result(nil);
  } else if ([@"Camera#releaseAllResources" isEqualToString:call.method]) {
    [self releaseAllResources];
    result(nil);
  } else {
    result(FlutterMethodNotImplemented);
  }
}

- (void)createCameraController:(FlutterMethodCall*)call {
  NSString *channelName = call.arguments[@"channelName"];
  FlutterMethodChannel* channel = [FlutterMethodChannel
                                methodChannelWithName:channelName
                                      binaryMessenger:[_registrar messenger]];

  NSString *cameraId = call.arguments[@"cameraId"];
  CameraController *controller = [[CameraController alloc] initWithCameraId:cameraId
                                                            textureRegistry:[_registrar textures]
                                                                  messenger:[_registrar messenger]];

  [_registrar addMethodCallDelegate:controller channel:channel];

  Pair *pair = [Pair new];
  pair.channel = channel;
  pair.controller = controller;
  [_controllers addObject:pair];
}

- (void)releaseAllResources {
  for (Pair *controllerPair in _controllers) {
    [controllerPair.controller close];
  }

  [_controllers removeAllObjects];
}
@end

File: super_camera/android/src/main/java/com/example/supercamera/SuperCameraPlugin.java
package com.example.supercamera;

import android.content.Context;
import android.hardware.camera2.CameraManager;
import android.os.Build;
import android.util.Pair;
import com.example.supercamera.base.BaseCameraController;
import com.example.supercamera.camera1.CameraController;
import com.example.supercamera.camera2.CameraController2;
import java.util.ArrayList;
import java.util.List;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** SuperCameraPlugin */
public class SuperCameraPlugin implements MethodCallHandler {
  private static final String PLUGIN_CHANNEL_NAME = "bmparr2450.plugins/super_camera";

  private final Registrar registrar;
  private final CameraManager cameraManager;
  private final List<Pair<MethodChannel, BaseCameraController>> controllers = new ArrayList<>();

  private SuperCameraPlugin(Registrar registrar, CameraManager manager) {
    this.registrar = registrar;
    this.cameraManager = manager;
  }

  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), PLUGIN_CHANNEL_NAME);

    if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
      final CameraManager manager =
          (CameraManager) registrar.activity().getSystemService(Context.CAMERA_SERVICE);
      channel.setMethodCallHandler(new SuperCameraPlugin(registrar, manager));
    } else {
      channel.setMethodCallHandler(new SuperCameraPlugin(registrar, null));
    }
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    switch(call.method) {
      case "Camera#availableCameras":
        availableCameras(result);
        break;
      case "Camera#createCameraController":
        createCameraController(call);
        result.success(null);
        break;
      case "Camera#releaseAllResources":
        releaseAllResources();
        result.success(null);
        break;
      default:
        result.notImplemented();
    }
  }

  private void availableCameras(Result result) {
    if (false && android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
      CameraController2.returnAvailableCameras(cameraManager, result);
    } else {
      result.success(CameraController.availableCameras());
    }
  }

  private void createCameraController(MethodCall call) {
    final String channelName = call.argument("channelName");
    final MethodChannel channel = new MethodChannel(registrar.messenger(), channelName);

    final String cameraId = call.argument("cameraId");

    final BaseCameraController controller;
    if (false && android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
      controller = new CameraController2(
          cameraId, registrar.textures(), registrar.messenger(), cameraManager);
    } else {
      controller = new CameraController(cameraId, registrar.textures(), registrar.messenger());
    }

    channel.setMethodCallHandler(controller);
    controllers.add(new Pair<>(channel, controller));
  }

  private void releaseAllResources() {
    for (Pair<MethodChannel, BaseCameraController> controllerPair : controllers) {
      controllerPair.second.close();
      controllerPair.first.setMethodCallHandler(null);
    }

    controllers.clear();
  }
}
